packages/flutter_test/lib/src/finders.dart

```diff
@@ -460,57 +469,544 @@ class CommonFinders {
+  /// specific type would provide 'of type MyObject', so that the example
+  /// sentence comes out to: "Found an item of type MyObject."
+  /// {@endtemplate}
+  ///
+  /// See also:
+  ///
+  ///   * [FinderBase.toString] where this is used to fully describe the finder
+  ///   * [FinderResult.toString] where this is used to provide context to the
+  ///     results of a search
+  String get requirements;
+
+  /// Returns all of the items that will be considered by this finder.
+  @protected
+  Iterable<TCandidate> get allCandidates;
+
+  /// Returns a variant of this finder that only matches the first item
+  /// found by this finder.
+  FinderBase<TCandidate> get first => _FirstFinder<TCandidate>(this);
+
+  /// Returns a variant of this finder that only matches the last item
+  /// found by this finder.
+  FinderBase<TCandidate> get last => _LastFinder<TCandidate>(this);
+
+  /// Returns a variant of this finder that only matches the item at the
+  /// given index found by this finder.
+  FinderBase<TCandidate> at(int index) => _IndexFinder<TCandidate>(this, index);
+
+  /// Returns all the items in the given list that match this
+  /// finder's requirements.
+  ///
+  /// When implementing your own finders that inherit directly from
+  /// [FinderBase], this is the main method to override. If your finder
   /// can efficiently be described just in terms of a predicate
-  /// function, consider extending [MatchFinder] instead.
-  Iterable<Element> apply(Iterable<Element> candidates);
+  /// function, consider mixing in [MatchFinderMixin] instead.
+  @protected
+  Iterable<TCandidate> findInCandidates(Iterable<TCandidate> candidates);
 
-  /// Whether this finder skips nodes that are offstage.
+  /// Searches all of the candidates and returns the result of that search.
   ///
-  /// If this is true, then the elements are walked using
+  /// See also:
+  ///
+  ///   * [found] which will return the latest results without re-executing the
+  ///     search.
+  ///   * [tryFind] which will indicate whether any results were found rather
+  ///     than directly returning results.
+  FinderResult<TCandidate> find() {
+    _found = FinderResult<TCandidate>(this.requirements, findInCandidates(allCandidates));
+    return found;
+  }
+
+  /// Searches all of the candidates and returns whether the search found any
+  /// matching candidates.
+  ///
+  /// This is useful in cases where an action needs to be repeated while or
+  /// until a finder has results. The results from the search can be accessed
+  /// using the [found] property without re-executing the search.
+  bool tryFind() {
+    find();
+    return found.isNotEmpty;
+  }
+
+  /// A string representation of this finder or its results.
+  ///
+  /// By default, this describes the results of the search in order to play
+  /// nicely with [expect] and its output when a failure occurs. If you wish
+  /// to get a string representation of the finder itself, pass [describeSelf]
```

Comment:
I wonder if this should just be a separate method?

Reply:
I'm conflicted on this method. The `matcher` library requires using `toString` to describe the results of a search, but `toString` is _supposed_ to describe the object itself. This was the compromise I came up with.

The thing I _like_ about this way of doing it is that when someone wants to call `toString` to debug or whatever, typing  that in immediately lets them know why the default behavior is what it is, and how to get the _expected_ `toString` behavior. We could also leave a breadcrumb to a different method, but this feels more correct to me given the limitations from `matcher`.